S3 support via a new adapter

Danielkonge opened the pull request:
This is still a draft and not completely ready yet, but I might need a bit of help with some of the non-S3 related things (see below). [Update: It is mostly ready now and I think it makes sense to get some feedback before I do much more on it.]

s3 is a bit different from ssh since we don't keep a constant connection and instead make new `aws s3 ls` calls when moving around, but I have still tried to build similar behavior to what is currently in `sshfs.lua`, see `s3fs.lua`.

Having the "filesystem" setup similarly, we then also build the adapter similar to what is done for ssh, see `ssh.lua` and `s3.lua`. (Again only with minor differences from the fact that we don't keep an active connection in the same way.)

I am still running into a few problems, among which the primary of the top of my head is:
- I had to temporarily use the scheme `oil-sthree://` since any scheme with a number in its name made it so that neovim didn't open the buffer correctly. I am not sure how to best work around this, but `oil-s3://` would be the clearly better scheme name here.
- The mutator is a bit hard to work with from the adapters perspective. It would probably be nicer to move some of its logic to each individual adapter instead.
- There are still a few minor bugs.
- I think there should be a better way to start the Oil s3 view from an already open neovim session, maybe just with `Oil --s3` or similar, but I haven't added that yet.
- Not really a problem, but more of a note: I haven't updated the README and docs yet. And I have a few `vim.notify` calls to track behavior that should of course be removed before this is ready.

For testing this out I made a new aws account and used `extra_s3_args = { "--profile=<my-profile>", "--region=<my-region>" }` when running stuff. 

If you want to edit some of this yourself if that is easier, feel free to edit directly on this branch. (Also let me know if you want cleaner commits, then I can just rebase.)

This PR will close #633.

Danielkonge commented on lua/oil/adapters/s3.lua:
```diff
@@ -0,0 +1,343 @@
+M.render_action = function(action)
+  if action.type == "create" then
+    local extra = action.type == "bucket" and "BUCKET " or ""
+    return string.format("CREATE %s%s", extra, action.url)
```
This was just because of a typo `type` instead of `entry_type`, it works fine now.

Danielkonge commented on lua/oil/init.lua:
```diff
@@ -1109,6 +1121,13 @@ local function close_preview_window_if_not_in_oil()
+local function escape_pattern(pat)
+  local res = pat:gsub("-", "%%-")
+  return res
+end
```
Reading other filetype patterns and inspecting `vim.filetype` a bit, I think we need to escape `-` in `oil-ssh` and `oil-s3`. (Just for the filetype patterns, not for the autocmd patterns.

Danielkonge replied:
Should I just remove this from this PR? It is unrelated to the s3 support.

stevearc commented:
If the region is only really needed for creating new buckets, then maybe we should just not allow the creation of buckets using oil. It seems like a pretty uncommon use case; I would imagine that most of the time this would be used for exploring and manipulating buckets that already exist.

If we also give up the special icon/highlight for buckets then I don't think we would need the new entry type at all. I know it's a slight downgrade in usability, but it keeps all the functionality and significantly decreases the complexity of this change. We might come up with some other way to customize the icons/highlights in the future, since this is something that people keep asking for.

Danielkonge commented:
> If the region is only really needed for creating new buckets, then maybe we should just not allow the creation of buckets using oil. It seems like a pretty uncommon use case; I would imagine that most of the time this would be used for exploring and manipulating buckets that already exist.

I think I might have confused a bit with what I wrote earlier. The region is generally not needed, and you can create buckets without the region. The only time you might run into problems is when you manually switch aws account with the `--profile` flag, but this is already an uncommon use case mostly for more advanced users, and they should then also be able to figure out how to add the region correctly. At most I think we might want to mention this in the docs, but I am not even sure that is worth it for such an edge case.

> If we also give up the special icon/highlight for buckets then I don't think we would need the new entry type at all. I know it's a slight downgrade in usability, but it keeps all the functionality and significantly decreases the complexity of this change. We might come up with some other way to customize the icons/highlights in the future, since this is something that people keep asking for.

I have removed the entry type and icon/highlight for buckets now, and everything should still work fine. 

I also updated the README and docs now.

stevearc commented on lua/oil/config.lua:
```diff
@@ -204,9 +206,14 @@ local default_config = {
+
+-- not "oil-s3://" on older neovim versions, since it doesn't open buffers correctly with a number
+-- in the name
+local oil_s3_string = vim.fn.has("nvim-0.12") and "oil-s3://" or "oil-sss://"
```
`has()` returns a 1 for true and 0 for false, which are both truthy in lua. You'll want to do `vim.fn.has("nvim-0.12") == 1 and ...`
